api/python/quilt3/packages.py

```diff
@@ -1141,24 +1141,30 @@ def _materialize(self, dest_url):
         pkg._meta = self._meta
         # Since all that is modified is physical keys, pkg will have the same top hash
         file_list = []
+        entries = []
         for logical_key, entry in self.walk():
             # Copy the datafiles in the package.
             physical_key = _to_singleton(entry.physical_keys)
+            unversioned_physical_key = physical_key.split('?', 1)[0]
```

Comment:
Can we split on `?versionId=` instead? I don't think we can safely assume that filenames won't have question marks in them.

Reply:
This is why we URL-encode paths...

Reply:
The logic we use does not work. Try
```
pkg = Package()
pkg.set("k", f"s3://quilt-ml-data/tst/myfile?.txt")
```
You will get a 404 despite the object existing. 

The core problem is illustrated by this function:
```
def test_question_mark():
    url = fix_url(str("s3://quilt-ml-data/tst/myfile?.txt"))
    parsed_url = urlparse(url)
    b, k, v = parse_s3_url(parsed_url)
    print("Expected:", "quilt-ml-data", "tst/myfile?.txt", None)
    print("Actual:", b, k, v)
```

We can fix it or not - `?` is [not a recommended character to use in an s3 key](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#object-keys), but I don't see a good reason to split on `?` instead of `?versionId=`

Reply:
Um... https://quilt-example.s3.amazonaws.com/dima/q/foo?.txt doesn't work, either - because it needs to be https://quilt-dima.s3.amazonaws.com/dima/q/foo%3F.txt . That is how URLs work. (Also, what if the file name contains "?versionId="?)

Reply:
I don't think this is worth blocking on, but there's currently no way for the user to know that they need to url encode the key that s3.list_object_v2 returns before calling pkg.set()